### PR TITLE
Migrate PluginManager and PackageManager to ES6

### PR DESCRIPTION
**Pull request by MrTimscampi**
**Changes**

Migrates pluginManager.js and packageManager.js to ES6, converts to classes and properly tags the main methods and properties as private.

Also adds a console group to the log messages we output when registering plugins, to make things a bit tidier.

**Issues**

### CODE REVIEW

**Comment by MrTimscampi** [comment]
I disabled the `no-unused-vars` rule temporarily in 757020b.

The new Babel parser for ESLint makes the stock `no-unused-vars` stumble on private class methods, which prevents linting. While waiting for ESLint to implement private class methods in their own rules or Babel to add an `@babel/no-unused-vars` rule to their plugin, the rule is turned off.

File: src/components/packageManager.js [hunk 1]
```diff
@@ -1,134 +1,137 @@
...
+            packageUrl += '/';
+            packageUrl += pluginUrl;
+
+            return packageUrl;
+        }
 
-        if (pkg) {
-            this.packagesList = this.packagesList.filter(function (p) {
-                return p.name !== name;
+        addPackage(packageManager, pkg) {
```
Comment (dmitrylyzo) [inline src/components/packageManager.js:72]: Isn't `packageManager === this`?

File: src/components/packageManager.js [hunk 2]
```diff
@@ -1,134 +1,137 @@
...
+            manifestUrls = manifestUrls.filter((i) => {
+                return i !== url;
+            });
+
+            appSettings.set(this.#settingsKey, JSON.stringify(manifestUrls));
         }
 
-        var packageUrl = pkg.url;
-        packageUrl = packageUrl.substring(0, packageUrl.lastIndexOf('/'));
+        loadPackage(packageManager, url, throwError) {
```
Comment (dmitrylyzo) [inline src/components/packageManager.js:90]: Isn't `packageManager === this`?

**Comment by dmitrylyzo** [comment]
https://github.com/jellyfin/jellyfin-web/blob/a609e3e2b4bc5dbca9b661f77e0a9d53a2222919/src/libraries/screensavermanager.js#L1
Still has quotes and `pluginManager = pluginManager.default || pluginManager;` is required.

File: src/scripts/site.js [hunk 3]
```diff
@@ -503,8 +503,9 @@ function initClient() {
 
         return new Promise(function (resolve, reject) {
             Promise.all(list.map(loadPlugin)).then(function () {
+                console.groupEnd('loading installed plugins');
```
[the remark this answers is not in the document]
Comment (MrTimscampi) [reply src/scripts/site.js:506]: It'd fit better in a `finally()`, probably

[the remark this answers is not in the document]
Comment (dmitrylyzo) [reply src/scripts/site.js:506]: But it must go before `packageManager`.
In fact, requirejs (at `loadPlugin`) just dies on fake plugin and doesn't call reject.

[the remark this answers is not in the document]
Comment (MrTimscampi) [reply src/scripts/site.js:506]: Yep, I changed it up to look like this:

```javascript
return new Promise(function (resolve, reject) {
    Promise.all(list.map(loadPlugin))
        .then(function () {
            console.debug('finished loading plugins');
        })
        .catch(() => {
            console.error(new Error('error while loading plugins'));
        })
        .finally(() => {
            console.groupEnd('loading installed plugins');
            import('packageManager').then(({default: packageManager}) => {
                console.groupCollapsed('loading packages');
                packageManager.init()
                    .then(resolve)
                    .catch(reject)
                    .finally(() => {
                        console.groupEnd('loading packages');
                    });
            });
        });
});
```

That should initialize packages even if pluginManager fails, and properly close the groups.

File: src/components/pluginManager.js [hunk 4]
```diff
@@ -49,103 +50,106 @@ define(['events', 'globalize'], function (events, globalize) {
             }
         }
 
-        if (typeof pluginSpec === 'string') {
-            console.debug('Loading plugin (via deprecated requirejs method): ' + pluginSpec);
+        loadPlugin(pluginSpec) {
+            var instance = this;
```
Comment (dmitrylyzo) [inline src/components/pluginManager.js:55]: unused
```suggestion
```

File: src/scripts/site.js [hunk 5]
```diff
@@ -470,13 +470,14 @@ function initClient() {
 
         require(['apphost', 'css!assets/css/librarybrowser'], function (appHost) {
             loadPlugins(appHost, browser).then(function () {
+                console.groupEnd('loading packages');
```
Comment (dmitrylyzo) [inline src/scripts/site.js:473]: leftover?

File: src/components/pluginManager.js [hunk 6]
```diff
@@ -49,103 +50,106 @@ define(['events', 'globalize'], function (events, globalize) {
...
-    PluginManager.prototype.mapPath = function (plugin, path, addCacheParam) {
-        if (typeof plugin === 'string') {
-            plugin = this.pluginsList.filter(function (p) {
-                return (p.id || p.packageName) === plugin;
-            })[0];
+            return '/plugins/' + plugin.id + '/' + route;
         }
 
-        var url = plugin.baseUrl + '/' + path;
+        #mapPath(plugin, path, addCacheParam) {
```
Comment (dmitrylyzo) [inline src/components/pluginManager.js:133]: This method is used here https://github.com/jellyfin/jellyfin-web/blob/e2aff662038bdf0fb730f9811975a02bb13a52be/src/plugins/logoScreensaver/plugin.js#L130

File: src/components/pluginManager.js [hunk 7]
```diff
@@ -1,37 +1,38 @@
...
+                strings: strings
+            });
+        }
 
-    function PluginManager() {
-        this.pluginsList = [];
-    }
+        #definePluginRoute(route, plugin) {
+            route.contentPath = this.#mapPath(plugin, route.path);
+            route.path = this.mapRoute(plugin, route);
```
Comment (dmitrylyzo) [inline src/components/pluginManager.js:25]: I can't get here
```suggestion
            route.path = this.#mapRoute(plugin, route);
```
or maybe just make it public?

File: src/scripts/site.js (hunk 5 above)
Comment (MrTimscampi) [reply src/scripts/site.js:473]: Indeed.

File: src/components/packageManager.js [hunk 8]
```diff
@@ -1,134 +1,141 @@
...
+
+    class PackageManager {
+        #packagesList = [];
+        #settingsKey = 'installedpackages1';
+
+        init() {
+            console.groupCollapsed('loading packages');
+            var manifestUrls = JSON.parse(appSettings.get(this.#settingsKey) || '[]');
+
+            var instance = this;
```
Comment (dmitrylyzo) [inline src/components/packageManager.js:13]: _SonarCloud_
```suggestion
```

File: src/components/packageManager.js [hunk 9]
```diff
@@ -1,134 +1,140 @@
...
+            packageUrl += pluginUrl;
+
+            return packageUrl;
+        }
 
-        if (pkg) {
-            this.packagesList = this.packagesList.filter(function (p) {
-                return p.name !== name;
+        addPackage(pkg) {
+            this.packagesList = this.packagesList.filter((p) => {
```
Comment (dmitrylyzo) [inline src/components/packageManager.js:76]:
```suggestion
            this.#packagesList = this.#packagesList.filter((p) => {
```

File: src/components/packageManager.js [hunk 10]
```diff
@@ -1,134 +1,140 @@
...
+                return p.name !== pkg.name;
             });
 
-            removeUrl(pkg.url);
+            this.packagesList.push(pkg);
```
Comment (dmitrylyzo) [inline src/components/packageManager.js:80]:
```suggestion
            this.#packagesList.push(pkg);
```

File: src/components/pluginManager.js [hunk 11]
```diff
@@ -49,103 +50,104 @@ define(['events', 'globalize'], function (events, globalize) {
...
 
-        var url = plugin.baseUrl + '/' + path;
+        mapPath(plugin, path, addCacheParam) {
+            if (typeof plugin === 'string') {
+                plugin = this.pluginsList.filter((p) => {
+                    return (p.id || p.packageName) === plugin;
+                })[0];
+            }
+
+            console.dir(plugin);
```
Comment (dmitrylyzo) [inline src/components/pluginManager.js:138]: Is this necessary?
It prints on screensaver activation.

**Pull request merged**